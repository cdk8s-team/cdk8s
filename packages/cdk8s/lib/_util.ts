export function removeEmpty(obj: any): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {
    return obj.map(x => removeEmpty(x));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = removeEmpty(value);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  return newObj;
}