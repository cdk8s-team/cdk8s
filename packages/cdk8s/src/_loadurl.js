// this is an executable program that downloads a file from a url (with
// redirects) and prints it's output to STDOUT. it is implemented as a program
// because node.js does not support issuing network requests synchronously and
// in the cdk we need everything to be synchronous (it all happens in the
// ctors). so we utilize `spawnSync` to spawn this program as a child process.
// alternatively we could have use `curl` but this is more portable.

const { http, https } = require('follow-redirects');
const { parse } = require('url');
const fs = require('fs');

const url = process.argv[2];
if (!url) {
  console.error(`Usage: ${process.argv[1]} URL`);
  process.exit(1);
}

const purl = parse(url);

if (!purl.protocol) {
  if (!purl.pathname) {
    throw new Error(`unable to parse pathname from file url: ${url}`);
  }

  process.stdout.write(fs.readFileSync(purl.pathname, 'utf-8'));
  process.exit(0);
}

const client = getHttpClient();

const get = client.get(url, response => {
  if (response.statusCode !== 200) {
    throw new Error(`${response.statusCode} response from http get: ${response.statusMessage}`);
  }

  response.on('data', chunk => process.stdout.write(chunk));
});

get.once('error', err => {
  throw new Error(`http error: ${err.message}`);
});

function getHttpClient() {
  switch (purl.protocol) {
    case 'http:': return http;
    case 'https:': return https;
    default:
      throw new Error(`unsupported protocol "${purl.protocol}" in url: ${url}`);
  }
}
