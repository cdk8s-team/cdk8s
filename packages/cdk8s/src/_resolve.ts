import { Lazy } from './lazy';

export function resolve(value: any): any {

  if (value == null) {
    return value;
  }

  // cdk8s token
  if (value instanceof Lazy) {
    const resolved = value.produce();
    return resolve(resolved);
  }

  // implicit token as generated by the cdk8s-cli (e.g IntOrString)
  if (typeof(value.resolve) === 'function') {
    const resolved = value.resolve();
    return resolve(resolved);
  }

  if (typeof(value) !== 'object') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(x => resolve(x));
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    result[k] = resolve(v);
  }

  return result;

}
