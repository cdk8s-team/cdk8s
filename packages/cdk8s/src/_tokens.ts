import { Tokenization, DefaultTokenResolver, StringConcat, Construct } from 'constructs';

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

export function resolve<T>(scope: Construct, obj: T): T {
  return Tokenization.resolve(obj, {
    scope,
    resolver: TOKEN_RESOLVER,
  });
}