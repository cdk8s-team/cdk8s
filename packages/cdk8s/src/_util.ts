export interface SanitizeOptions {
  /**
   * Do not include empty objects (no keys).
   */
  readonly filterEmptyObjects?: boolean;

  /**
   * Do not include arrays with no items.
   */
  readonly filterEmptyArrays?: boolean;
}

export function sanitizeValue(obj: any, options: SanitizeOptions = { }): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {

    if (options.filterEmptyArrays && obj.length === 0) {
      return undefined;
    }

    return obj.map(x => sanitizeValue(x, options));
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = sanitizeValue(value, options);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  if (options.filterEmptyObjects && Object.keys(newObj).length === 0) {
    return undefined;
  }

  return newObj;
}
