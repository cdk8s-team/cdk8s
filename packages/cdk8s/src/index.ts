export * from './api-object';
export * from './chart';
export * from './testing';
export * from './app';
export * from './include';
export * from './yaml';
export * from './metadata';