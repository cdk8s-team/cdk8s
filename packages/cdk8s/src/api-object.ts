import { Construct, IConstruct, Node } from 'constructs';
import { Chart } from './chart';
import { sanitizeValue } from './_util';
import { resolve } from './_tokens';
import * as stringify from 'json-stable-stringify';
import { ApiObjectMetadata, ApiObjectMetadataDefinition } from './metadata';

/**
 * Options for defining API objects.
 */
export interface ApiObjectOptions {
  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ApiObjectMetadata;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

export class ApiObject extends Construct {
  /**
   * The name of the API object.
   *
   * If a name is specified in `metadata.name` this will be the name returned.
   * Otherwise, a name will be generated by calling
   * `Chart.of(this).generatedObjectName(this)`, which by default uses the
   * construct path to generate a DNS-compatible name for the resource.
   */
  public readonly name: string;

  /**
   * The object's API version.
   */
  public readonly apiVersion: string;

  /**
   * The object kind.
   */
  public readonly kind: string;

  /**
   * The chart in which this object is defined.
   */
  public readonly chart: Chart;

  /**
   * Metadata associated with this API object.
   */
  public readonly metadata: ApiObjectMetadataDefinition;

  /**
   * Defines an API object.
   *
   * @param scope the construct scope
   * @param ns namespace
   * @param options options
   */
  constructor(scope: Construct, ns: string, private readonly options: ApiObjectOptions) {
    super(scope, ns);
    this.chart = Chart.of(this);
    this.kind = options.kind;
    this.apiVersion = options.apiVersion;

    this.name = options.metadata?.name ?? this.chart.generateObjectName(this);

    this.metadata = new ApiObjectMetadataDefinition({
      name: this.name,
      namespace: options.metadata?.namespace ?? this.chart.namespace,

      // override user defined values
      ...options.metadata,
    });
  }

  /**
   * Create a dependency between this ApiObject and other constructs.
   * These can be other ApiObjects, Charts, or custom.
   *
   * @param dependencies the dependencies to add.
   */
  public addDependency(...dependencies: IConstruct[]) {
    Node.of(this).addDependency(...dependencies);
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const data = {
      ...this.options,
      metadata: this.metadata.toJson(),
    };

    // convert to "pure data" so, for example, when we convert to yaml these
    // references are not converted to anchors.
    return JSON.parse(stringify(sanitizeValue(resolve(this, data))));
  }
}
