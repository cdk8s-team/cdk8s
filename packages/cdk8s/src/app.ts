import * as fs from 'fs';
import * as path from 'path';
import { Construct, Node, IConstruct } from 'constructs';
import { ApiObject } from './api-object';
import { Chart } from './chart';
import { DependencyGraph } from './dependency';
import { Names } from './names';
import { Yaml } from './yaml';

export interface AppProps {
  /**
   * The directory to output Kubernetes manifests.
   *
   * @default - CDK8S_OUTDIR if defined, otherwise "dist"
   */
  readonly outdir?: string;
}

/**
 * Represents a cdk8s application.
 */
export class App extends Construct {
  /**
   * Synthesize a single chart.
   *
   * Each element returned in the resulting array represents a different ApiObject
   * in the scope of the chart.
   *
   * Note that the returned array order is important. It is determined by the various dependencies between
   * the constructs in the chart, where the first element is the one without dependencies, and so on...
   *
   * @returns An array of JSON objects.
   * @param chart the chart to synthesize.
   * @internal
   */
  public static _synthChart(chart: Chart): any[] {

    const app: App = App.of(chart);

    // we must prepare the entire app before synthesizing the chart
    // because the dependency inference happens on the app level.
    resolveDependencies(app);

    // validate the app since we want to call onValidate of the relevant constructs.
    // note this will also call onValidate on constructs from possibly different charts,
    // but thats ok too since we no longer treat constructs as a self-contained synthesis unit.
    validate(app);

    return chartToKube(chart);
  }

  private static of(c: IConstruct): App {

    const scope = Node.of(c).scope;

    if (!scope) {
      // the app is the only construct without a scope.
      return c as App;
    }

    return App.of(scope);
  }

  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /**
   * Defines an app
   * @param props configuration options
   */
  constructor(props: AppProps = { }) {
    super(undefined as any, '');
    this.outdir = props.outdir ?? process.env.CDK8S_OUTDIR ?? 'dist';
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {

    fs.mkdirSync(this.outdir, { recursive: true });

    // this is kind of sucky, eventually I would like the DependencyGraph
    // to be able to answer this question.
    const hasDependantCharts = resolveDependencies(this);

    // Since we plan on removing the distributed synth mechanism, we no longer call `Node.synthesize`, but rather simply implement
    // the necessary operations. We do however want to preserve the distributed validation.
    validate(this);

    const simpleManifestNamer = (chart: Chart) => `${Names.toDnsLabel(Node.of(chart).path)}.k8s.yaml`;
    const manifestNamer = hasDependantCharts ? (chart: Chart) => `${index.toString().padStart(4, '0')}-${simpleManifestNamer(chart)}` : simpleManifestNamer;

    const charts: IConstruct[] = new DependencyGraph(Node.of(this)).topology().filter(x => x instanceof Chart);

    let index = 0;
    for (const node of charts) {
      const chart: Chart = Chart.of(node);
      Yaml.save(path.join(this.outdir, manifestNamer(chart)), chartToKube(chart));
      index++;
    }

  }
}

function validate(app: App) {

  // Note this is a copy-paste of https://github.com/aws/constructs/blob/master/lib/construct.ts#L438.
  const errors = Node.of(app).validate();
  if (errors.length > 0) {
    const errorList = errors.map(e => `[${Node.of(e.source).path}] ${e.message}`).join('\n  ');
    throw new Error(`Validation failed with the following errors:\n  ${errorList}`);
  }

}

function resolveDependencies(app: App) {

  let hasDependantCharts = false;

  for (const dep of Node.of(app).dependencies) {

    // create explicit api object dependencies from implicit construct dependencies
    const targetApiObjects = Node.of(dep.target).findAll().filter(c => c instanceof ApiObject);
    const sourceApiObjects = Node.of(dep.source).findAll().filter(c => c instanceof ApiObject);

    for (const target of targetApiObjects) {
      for (const source of sourceApiObjects) {
        Node.of(source).addDependency(target);
      }
    }

    // create an explicit chart dependency from implicit construct dependencies
    const sourceChart = Chart.of(dep.source);
    const targetChart = Chart.of(dep.target);

    if (sourceChart !== targetChart) {
      Node.of(sourceChart).addDependency(targetChart);
      hasDependantCharts = true;
    }

  }

  return hasDependantCharts;

}

function chartToKube(chart: Chart) {
  return new DependencyGraph(Node.of(chart)).topology()
    .filter(x => x instanceof ApiObject)
    .map(x => (x as ApiObject).toJson());
}
