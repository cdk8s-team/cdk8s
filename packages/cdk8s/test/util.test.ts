import { sanitizeValue } from '../src/_util';

describe('sanitizeValue', () => {

  test('default options', () => {
    expect(sanitizeValue(null)).toBe(undefined);
    expect(sanitizeValue(undefined)).toBe(undefined);
    expect(sanitizeValue({ })).toStrictEqual({ });
    expect(sanitizeValue([])).toStrictEqual([]);
    expect(sanitizeValue(1)).toBe(1);
    expect(sanitizeValue({ hello: 123 })).toStrictEqual({ hello: 123 });
    expect(sanitizeValue([1, 2, 3])).toStrictEqual([1, 2, 3]);
    expect(sanitizeValue({ xoo: 123, foo: [] })).toStrictEqual({ xoo: 123, foo: [] });
    expect(sanitizeValue({ xoo: { }, foo: { bar: { zoo: undefined, hey: { }, me: 123 } } }))
      .toStrictEqual({ xoo: { }, foo: { bar: { hey: { }, me: 123 } } });
    expect(sanitizeValue({ xoo: 123, foo: [1, 2, { foo: 123, bar: undefined, zoo: [] }, 3] }))
      .toStrictEqual({ xoo: 123, foo: [1, 2, { foo: 123, zoo: [] }, 3] });
    expect(sanitizeValue([1, 2, 3, [], { }, 4])).toStrictEqual([1, 2, 3, [], { }, 4]); // special case

    expect(() => sanitizeValue(new Dummy())).toThrow(/can't render non-simple object of type 'Dummy'/);
  });

  test('filterEmptyArrays', () => {
    const options = {
      filterEmptyArrays: true,
    };

    expect(sanitizeValue([], options)).toBe(undefined);
    expect(sanitizeValue({ foo: [], bar: [1, 2] }, options)).toStrictEqual({ bar: [1, 2] });
    expect(sanitizeValue({ foo: { bar: [] } }, options)).toStrictEqual({ foo: { } });
  });

  test('filterEmptyObjects', () => {
    const options = {
      filterEmptyObjects: true,
    };

    expect(sanitizeValue({ }, options)).toBe(undefined);
    expect(sanitizeValue({ foo: { }, bar: { hey: 'there' } }, options)).toStrictEqual({ bar: { hey: 'there' } });
    expect(sanitizeValue({ foo: { bar: { } } }, options)).toStrictEqual(undefined);
  });

  test('sortKeys', () => {
    const input = { zzz: 999, aaa: 111, nested: { foo: { zag: [1, 2, 3], bar: '1111' } } };
    expect(str(sanitizeValue(input))).toMatchSnapshot();
    expect(str(sanitizeValue(input, { sortKeys: false }))).toMatchSnapshot();
  });

  class Dummy { }
});

function str(obj: any) {
  return JSON.stringify(obj, undefined, 2);
}