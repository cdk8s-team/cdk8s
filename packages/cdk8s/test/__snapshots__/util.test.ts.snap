// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitizeValue sortKeys 1`] = `
"{
  \\"aaa\\": 111,
  \\"nested\\": {
    \\"foo\\": {
      \\"bar\\": \\"1111\\",
      \\"zag\\": [
        1,
        2,
        3
      ]
    }
  },
  \\"zzz\\": 999
}"
`;

exports[`sanitizeValue sortKeys 2`] = `
"{
  \\"zzz\\": 999,
  \\"aaa\\": 111,
  \\"nested\\": {
    \\"foo\\": {
      \\"zag\\": [
        1,
        2,
        3
      ],
      \\"bar\\": \\"1111\\"
    }
  }
}"
`;
